File: yasmin/test/test_state_machine.cpp
// Copyright (C) 2023 Miguel Ángel González Santamarta
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include <gtest/gtest.h>
#include <memory>
#include <string>

#include "yasmin/blackboard.hpp"
#include "yasmin/state.hpp"
#include "yasmin/state_machine.hpp"
#include "yasmin/types.hpp"

using namespace yasmin;

class FooState : public State {
private:
  int counter;

public:
  FooState() : State({"outcome1", "outcome2"}), counter(0) {}

  std::string execute(yasmin::Blackboard::SharedPtr blackboard) override {
    if (counter < 3) {
      counter++;
      blackboard->set<std::string>("foo_str",
                                   "Counter: " + std::to_string(counter));
      return "outcome1";
    } else {
      return "outcome2";
    }
  }
};

class BarState : public State {
public:
  BarState() : State({"outcome2", "outcome3"}) {}

  std::string execute(yasmin::Blackboard::SharedPtr blackboard) override {
    return "outcome2";
  }
};

class TestStateMachine : public ::testing::Test {
protected:
  yasmin::StateMachine::SharedPtr sm;
  yasmin::Blackboard::SharedPtr blackboard;

  void SetUp() override {
    sm = StateMachine::make_shared(yasmin::Outcomes{"outcome4", "outcome5"});
    blackboard = yasmin::Blackboard::make_shared();

    sm->add_state(
        "FOO", std::make_shared<FooState>(),
        yasmin::Transitions{{"outcome1", "BAR"}, {"outcome2", "outcome4"}});

    sm->add_state("BAR", std::make_shared<BarState>(),
                  yasmin::Transitions{{"outcome2", "FOO"}});
  }
};

TEST_F(TestStateMachine, TestStr) {
  std::string sm_str = sm->to_string();
  // Check if "Bar (BarState)" and "Foo (FooState)" are in the string
  EXPECT_TRUE(sm_str.find("BAR (BarState)") != std::string::npos);
  EXPECT_TRUE(sm_str.find("FOO (FooState)") != std::string::npos);
}

TEST_F(TestStateMachine, TestGetNameEmpty) { EXPECT_EQ(sm->get_name(), ""); }

TEST_F(TestStateMachine, TestSetName) {
  sm->set_name("TestSM");
  EXPECT_EQ(sm->get_name(), "TestSM");
}

TEST_F(TestStateMachine, TestGetStates) {
  auto states = sm->get_states();
  EXPECT_EQ(states.size(), 2);
  EXPECT_TRUE(states.find("FOO") != states.end());
  EXPECT_TRUE(states.find("BAR") != states.end());
}

TEST_F(TestStateMachine, TestGetStartState) {
  EXPECT_EQ(sm->get_start_state(), "FOO");
  sm->set_start_state("BAR");
  EXPECT_EQ(sm->get_start_state(), "BAR");
}

TEST_F(TestStateMachine, TestGetCurrentState) {
  EXPECT_EQ(sm->get_current_state(), "");
}

TEST_F(TestStateMachine, TestStateCall) {
  EXPECT_EQ((*sm)(blackboard), "outcome4");
}

TEST_F(TestStateMachine, TestSetStartStateEmpty) {
  try {
    sm->set_start_state("");
    FAIL() << "Expected std::invalid_argument";
  } catch (const std::invalid_argument &e) {
    EXPECT_STREQ("Initial state cannot be empty", e.what());
  }
}

TEST_F(TestStateMachine, TestSetStartStateWrongState) {
  try {
    sm->set_start_state("FOO1");
    FAIL() << "Expected std::invalid_argument";
  } catch (const std::invalid_argument &e) {
    EXPECT_STREQ("Initial state 'FOO1' is not in the state machine", e.what());
  }
}

TEST_F(TestStateMachine, TestAddRepeatedState) {
  try {
    sm->add_state("FOO", std::make_shared<FooState>(),
                  yasmin::Transitions{{"outcome1", "BAR"}});
    FAIL() << "Expected std::logic_error";
  } catch (const std::logic_error &e) {
    EXPECT_STREQ("State 'FOO' already registered in the state machine",
                 e.what());
  }
}

TEST_F(TestStateMachine, TestAddOutcomeState) {
  try {
    sm->add_state("outcome4", std::make_shared<FooState>(),
                  yasmin::Transitions{{"outcome1", "BAR"}});
    FAIL() << "Expected std::logic_error";
  } catch (const std::logic_error &e) {
    EXPECT_STREQ("State name 'outcome4' is already registered as an outcome",
                 e.what());
  }
}

TEST_F(TestStateMachine, TestAddStateWithWrongOutcome) {
  try {
    sm->add_state("FOO1", std::make_shared<FooState>(),
                  yasmin::Transitions{{"outcome9", "BAR"}});
    FAIL() << "Expected std::invalid_argument";
  } catch (const std::invalid_argument &e) {
    std::string error_msg = e.what();
    EXPECT_TRUE(error_msg.find("State 'FOO1' references unregistered outcomes "
                               "'outcome9'") != std::string::npos);
    EXPECT_TRUE(error_msg.find("available outcomes are") != std::string::npos);
  }
}

TEST_F(TestStateMachine, TestAddWrongSourceTransition) {
  try {
    sm->add_state("FOO1", std::make_shared<FooState>(),
                  yasmin::Transitions{{"", "BAR"}});
    FAIL() << "Expected std::invalid_argument";
  } catch (const std::invalid_argument &e) {
    EXPECT_STREQ("Transitions with empty source in state 'FOO1'", e.what());
  }
}

TEST_F(TestStateMachine, TestAddWrongTargetTransition) {
  try {
    sm->add_state("FOO1", std::make_shared<FooState>(),
                  yasmin::Transitions{{"outcome1", ""}});
    FAIL() << "Expected std::invalid_argument";
  } catch (const std::invalid_argument &e) {
    EXPECT_STREQ("Transitions with empty target in state 'FOO1'", e.what());
  }
}

TEST_F(TestStateMachine, TestValidateOutcomeFromFsmNotUsed) {
  auto sm1 = StateMachine::make_shared(yasmin::Outcomes{"outcome4"});
  auto sm2 =
      StateMachine::make_shared(yasmin::Outcomes{"outcome4", "outcome5"});
  sm1->add_state("FSM", sm2);
  sm2->add_state("FOO", std::make_shared<FooState>(),
                 {
                     {"outcome1", "outcome4"},
                     {"outcome2", "outcome4"},
                 });

  try {
    sm1->validate(true);
    FAIL() << "Expected std::runtime_error";
  } catch (const std::runtime_error &e) {
    EXPECT_STREQ("State 'FSM' outcome 'outcome5' not registered in transitions",
                 e.what());
  }
}

TEST_F(TestStateMachine, TestValidateOutcomeFromStateNotUsed) {
  auto sm1 = StateMachine::make_shared(yasmin::Outcomes{"outcome4"});
  auto sm2 = StateMachine::make_shared(yasmin::Outcomes{"outcome4"});
  sm1->add_state("FSM", sm2);
  sm2->add_state("FOO", std::make_shared<FooState>(),
                 {
                     {"outcome1", "outcome4"},
                 });

  try {
    sm1->validate(true);
    FAIL() << "Expected std::runtime_error";
  } catch (const std::runtime_error &e) {
    EXPECT_STREQ("State 'FOO' outcome 'outcome2' not registered in transitions",
                 e.what());
  }
}

TEST_F(TestStateMachine, TestValidateFsmOutcomeNotUsed) {
  auto sm1 = StateMachine::make_shared(yasmin::Outcomes{"outcome4"});
  auto sm2 =
      StateMachine::make_shared(yasmin::Outcomes{"outcome4", "outcome5"});
  sm1->add_state("FSM", sm2,
                 {
                     {"outcome5", "outcome4"},
                 });
  sm2->add_state("FOO", std::make_shared<FooState>(),
                 {
                     {"outcome1", "outcome4"},
                     {"outcome2", "outcome4"},
                 });

  try {
    sm1->validate(true);
    FAIL() << "Expected std::runtime_error";
  } catch (const std::runtime_error &e) {
    EXPECT_STREQ("Target outcome 'outcome5' not registered in transitions",
                 e.what());
  }
}

TEST_F(TestStateMachine, TestValidateWrongState) {
  auto sm1 = StateMachine::make_shared(yasmin::Outcomes{"outcome4"});
  auto sm2 = StateMachine::make_shared(yasmin::Outcomes{"outcome4"});
  sm1->add_state("FSM", sm2,
                 {
                     {"outcome4", "outcome4"},
                 });
  sm2->add_state("FOO", std::make_shared<FooState>(),
                 {
                     {"outcome1", "BAR"},
                     {"outcome2", "outcome4"},
                 });

  try {
    sm1->validate();
    FAIL() << "Expected std::runtime_error";
  } catch (const std::runtime_error &e) {
    EXPECT_STREQ(
        "State machine outcome 'BAR' not registered as outcome neither state",
        e.what());
  }
}

TEST_F(TestStateMachine, TestStartCallback) {
  bool start_called = false;
  std::string start_state_called;

  sm->add_start_cb([&](yasmin::Blackboard::SharedPtr blackboard,
                       const std::string &start_state) {
    start_called = true;
    start_state_called = start_state;
  });

  EXPECT_EQ((*sm)(blackboard), "outcome4");
  EXPECT_TRUE(start_called);
  EXPECT_EQ(start_state_called, "FOO");
}

TEST_F(TestStateMachine, TestTransitionCallback) {
  std::vector<std::tuple<std::string, std::string, std::string>>
      transitions_called;

  sm->add_transition_cb([&](yasmin::Blackboard::SharedPtr blackboard,
                            const std::string &from_state,
                            const std::string &to_state,
                            const std::string &outcome) {
    transitions_called.emplace_back(from_state, to_state, outcome);
  });

  EXPECT_EQ((*sm)(blackboard), "outcome4");

  // Should have transitions: FOO -> BAR (outcome1), BAR -> FOO (outcome2), FOO
  // -> BAR (outcome1), BAR -> FOO (outcome2), FOO -> BAR (outcome1), BAR -> FOO
  // (outcome2)
  EXPECT_EQ(transitions_called.size(), 6);
  EXPECT_EQ(std::get<0>(transitions_called[0]), "FOO");
  EXPECT_EQ(std::get<1>(transitions_called[0]), "BAR");
  EXPECT_EQ(std::get<2>(transitions_called[0]), "outcome1");
  EXPECT_EQ(std::get<0>(transitions_called[1]), "BAR");
  EXPECT_EQ(std::get<1>(transitions_called[1]), "FOO");
  EXPECT_EQ(std::get<2>(transitions_called[1]), "outcome2");
  EXPECT_EQ(std::get<0>(transitions_called[2]), "FOO");
  EXPECT_EQ(std::get<1>(transitions_called[2]), "BAR");
  EXPECT_EQ(std::get<2>(transitions_called[2]), "outcome1");
  EXPECT_EQ(std::get<0>(transitions_called[3]), "BAR");
  EXPECT_EQ(std::get<1>(transitions_called[3]), "FOO");
  EXPECT_EQ(std::get<2>(transitions_called[3]), "outcome2");
  EXPECT_EQ(std::get<0>(transitions_called[4]), "FOO");
  EXPECT_EQ(std::get<1>(transitions_called[4]), "BAR");
  EXPECT_EQ(std::get<2>(transitions_called[4]), "outcome1");
  EXPECT_EQ(std::get<0>(transitions_called[5]), "BAR");
  EXPECT_EQ(std::get<1>(transitions_called[5]), "FOO");
  EXPECT_EQ(std::get<2>(transitions_called[5]), "outcome2");
}

TEST_F(TestStateMachine, TestEndCallback) {
  bool end_called = false;
  std::string end_outcome_called;

  sm->add_end_cb([&](yasmin::Blackboard::SharedPtr blackboard,
                     const std::string &outcome) {
    end_called = true;
    end_outcome_called = outcome;
  });

  EXPECT_EQ((*sm)(blackboard), "outcome4");
  EXPECT_TRUE(end_called);
  EXPECT_EQ(end_outcome_called, "outcome4");
}

TEST_F(TestStateMachine, TestMultipleCallbacks) {
  int start_count = 0;
  int transition_count = 0;
  int end_count = 0;

  sm->add_start_cb([&](yasmin::Blackboard::SharedPtr, const std::string &) {
    start_count++;
  });
  sm->add_start_cb([&](yasmin::Blackboard::SharedPtr, const std::string &) {
    start_count++;
  });

  sm->add_transition_cb([&](yasmin::Blackboard::SharedPtr, const std::string &,
                            const std::string &,
                            const std::string &) { transition_count++; });
  sm->add_transition_cb([&](yasmin::Blackboard::SharedPtr, const std::string &,
                            const std::string &,
                            const std::string &) { transition_count++; });

  sm->add_end_cb(
      [&](yasmin::Blackboard::SharedPtr, const std::string &) { end_count++; });
  sm->add_end_cb(
      [&](yasmin::Blackboard::SharedPtr, const std::string &) { end_count++; });

  EXPECT_EQ((*sm)(blackboard), "outcome4");
  EXPECT_EQ(start_count, 2);
  EXPECT_EQ(transition_count, 12); // 6 transitions * 2 callbacks
  EXPECT_EQ(end_count, 2);
}

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: yasmin/test/test_state_machine.py
# Copyright (C) 2023 Miguel Ángel González Santamarta

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


import threading
import time
import unittest
from yasmin import Blackboard, StateMachine, State


class FooState(State):
    def __init__(self):
        super().__init__(["outcome1", "outcome2"])
        self.counter = 0

    def execute(self, blackboard):
        if self.counter < 3:
            self.counter += 1
            blackboard["foo_str"] = f"Counter: {self.counter}"
            return "outcome1"

        else:
            return "outcome2"


class BarState(State):
    def __init__(self):
        super().__init__(outcomes=["outcome2", "outcome3"])

    def execute(self, blackboard):
        return "outcome2"


class TestStateMachine(unittest.TestCase):

    maxDiff = None

    def setUp(self):
        self.sm = StateMachine(outcomes=["outcome4", "outcome5"])

        self.sm.add_state(
            "FOO",
            FooState(),
            transitions={
                "outcome1": "BAR",
                "outcome2": "outcome4",
            },
        )
        self.sm.add_state(
            "BAR",
            BarState(),
            transitions={"outcome2": "FOO"},
        )

    def test_str(self):
        string = str(self.sm)
        # Check if "Bar (BarState)" and "Foo (FooState)" are in the string
        self.assertIn("BAR (BarState)", string)
        self.assertIn("FOO (FooState)", string)

    def test_get_name_empty(self):
        self.assertEqual("", self.sm.get_name())

    def test_set_name(self):
        self.sm.set_name("TestSM")
        self.assertEqual("TestSM", self.sm.get_name())

    def test_get_states(self):
        self.assertTrue(isinstance(self.sm.get_states()["FOO"]["state"], FooState))
        self.assertTrue(isinstance(self.sm.get_states()["BAR"]["state"], BarState))

    def test_get_start_state(self):
        self.assertEqual("FOO", self.sm.get_start_state())
        self.sm.set_start_state("BAR")
        self.assertEqual("BAR", self.sm.get_start_state())

    def test_get_current_state(self):
        self.assertEqual("", self.sm.get_current_state())

    def test_state_call(self):
        self.assertEqual("outcome4", self.sm())

    def test_set_start_state_empty(self):
        with self.assertRaises(ValueError) as context:
            self.sm.set_start_state("")
        self.assertEqual(str(context.exception), "Initial state cannot be empty")

    def test_set_start_state_wrong_state(self):
        with self.assertRaises(ValueError) as context:
            self.sm.set_start_state("FOO1")
        self.assertEqual(
            str(context.exception),
            "Initial state 'FOO1' is not in the state machine",
        )

    def test_add_repeated_state(self):
        with self.assertRaises(RuntimeError) as context:
            self.sm.add_state(
                "FOO",
                FooState(),
                transitions={
                    "outcome1": "BAR",
                },
            )
        self.assertEqual(
            str(context.exception),
            "State 'FOO' already registered in the state machine",
        )

    def test_add_outcome_state(self):
        with self.assertRaises(RuntimeError) as context:
            self.sm.add_state(
                "outcome4",
                FooState(),
                transitions={
                    "outcome1": "BAR",
                },
            )
        self.assertEqual(
            str(context.exception),
            "State name 'outcome4' is already registered as an outcome",
        )

    def test_add_state_with_wrong_outcome(self):
        with self.assertRaises(ValueError) as context:
            self.sm.add_state(
                "FOO1",
                FooState(),
                transitions={
                    "outcome9": "BAR",
                },
            )
        self.assertEqual(
            str(context.exception),
            "State 'FOO1' references unregistered outcomes 'outcome9', available outcomes are ['outcome1', 'outcome2']",
        )

    def test_add_wrong_source_transition(self):
        with self.assertRaises(ValueError) as context:
            self.sm.add_state(
                "FOO1",
                FooState(),
                transitions={
                    "": "BAR",
                },
            )
        self.assertEqual(
            str(context.exception),
            "Transitions with empty source in state 'FOO1'",
        )

    def test_add_wrong_target_transition(self):
        with self.assertRaises(ValueError) as context:
            self.sm.add_state(
                "FOO1",
                FooState(),
                transitions={
                    "outcome1": "",
                },
            )
        self.assertEqual(
            str(context.exception), "Transitions with empty target in state 'FOO1'"
        )

    def test_validate_outcome_from_fsm_not_used(self):

        sm_1 = StateMachine(outcomes=["outcome4"])

        sm_2 = StateMachine(outcomes=["outcome4", "outcome5"])
        sm_1.add_state("FSM", sm_2)

        sm_2.add_state(
            "FOO",
            FooState(),
            transitions={
                "outcome1": "outcome4",
                "outcome2": "outcome4",
            },
        )

        with self.assertRaises(RuntimeError) as context:
            sm_1.validate(True)
        self.assertEqual(
            str(context.exception),
            "State 'FSM' outcome 'outcome5' not registered in transitions",
        )

    def test_validate_outcome_from_state_not_used(self):

        sm_1 = StateMachine(outcomes=["outcome4"])

        sm_2 = StateMachine(outcomes=["outcome4"])
        sm_1.add_state("FSM", sm_2)

        sm_2.add_state(
            "FOO",
            FooState(),
            transitions={
                "outcome1": "outcome4",
            },
        )

        with self.assertRaises(RuntimeError) as context:
            sm_1.validate(True)
        self.assertEqual(
            str(context.exception),
            "State 'FOO' outcome 'outcome2' not registered in transitions",
        )

    def test_validate_fsm_outcome_not_used(self):

        sm_1 = StateMachine(outcomes=["outcome4"])

        sm_2 = StateMachine(outcomes=["outcome4", "outcome5"])
        sm_1.add_state(
            "FSM",
            sm_2,
            transitions={
                "outcome5": "outcome4",
            },
        )

        sm_2.add_state(
            "FOO",
            FooState(),
            transitions={
                "outcome1": "outcome4",
                "outcome2": "outcome4",
            },
        )

        with self.assertRaises(RuntimeError) as context:
            sm_1.validate(True)
        self.assertEqual(
            str(context.exception),
            "Target outcome 'outcome5' not registered in transitions",
        )

    def test_validate_wrong_state(self):

        sm_1 = StateMachine(outcomes=["outcome4"])

        sm_2 = StateMachine(outcomes=["outcome4"])
        sm_1.add_state(
            "FSM",
            sm_2,
            transitions={
                "outcome4": "outcome4",
            },
        )

        sm_2.add_state(
            "FOO",
            FooState(),
            transitions={
                "outcome1": "BAR",
                "outcome2": "outcome4",
            },
        )

        with self.assertRaises(RuntimeError) as context:
            sm_1.validate()
        self.assertEqual(
            str(context.exception),
            "State machine outcome 'BAR' not registered as outcome neither state",
        )

    def test_start_callback(self):
        start_called = False
        start_state_called = None

        def start_cb(blackboard, start_state):
            nonlocal start_called, start_state_called
            start_called = True
            start_state_called = start_state

        self.sm.add_start_cb(start_cb)
        result = self.sm()
        self.assertEqual(result, "outcome4")
        self.assertTrue(start_called)
        self.assertEqual(start_state_called, "FOO")

    def test_transition_callback(self):
        transitions_called = []

        def transition_cb(blackboard, from_state, to_state, outcome):
            transitions_called.append((from_state, to_state, outcome))

        self.sm.add_transition_cb(transition_cb)
        result = self.sm()
        self.assertEqual(result, "outcome4")

        # Should have transitions: FOO -> BAR (outcome1), BAR -> FOO (outcome2), FOO -> BAR (outcome1), BAR -> FOO (outcome2), FOO -> BAR (outcome1), BAR -> FOO (outcome2)
        self.assertEqual(len(transitions_called), 6)
        self.assertEqual(transitions_called[0], ("FOO", "BAR", "outcome1"))
        self.assertEqual(transitions_called[1], ("BAR", "FOO", "outcome2"))
        self.assertEqual(transitions_called[2], ("FOO", "BAR", "outcome1"))
        self.assertEqual(transitions_called[3], ("BAR", "FOO", "outcome2"))
        self.assertEqual(transitions_called[4], ("FOO", "BAR", "outcome1"))
        self.assertEqual(transitions_called[5], ("BAR", "FOO", "outcome2"))

    def test_end_callback(self):
        end_called = False
        end_outcome_called = None

        def end_cb(blackboard, outcome):
            nonlocal end_called, end_outcome_called
            end_called = True
            end_outcome_called = outcome

        self.sm.add_end_cb(end_cb)
        result = self.sm()
        self.assertEqual(result, "outcome4")
        self.assertTrue(end_called)
        self.assertEqual(end_outcome_called, "outcome4")

    def test_multiple_callbacks(self):
        start_count = [0]
        transition_count = [0]
        end_count = [0]

        self.sm.add_start_cb(
            lambda blackboard, start_state: start_count.__setitem__(0, start_count[0] + 1)
        )
        self.sm.add_start_cb(
            lambda blackboard, start_state: start_count.__setitem__(0, start_count[0] + 1)
        )

        self.sm.add_transition_cb(
            lambda blackboard, from_state, to_state, outcome: transition_count.__setitem__(
                0, transition_count[0] + 1
            )
        )
        self.sm.add_transition_cb(
            lambda blackboard, from_state, to_state, outcome: transition_count.__setitem__(
                0, transition_count[0] + 1
            )
        )

        self.sm.add_end_cb(
            lambda blackboard, outcome: end_count.__setitem__(0, end_count[0] + 1)
        )
        self.sm.add_end_cb(
            lambda blackboard, outcome: end_count.__setitem__(0, end_count[0] + 1)
        )

        result = self.sm()
        self.assertEqual(result, "outcome4")
        self.assertEqual(start_count[0], 2)
        self.assertEqual(transition_count[0], 12)
        self.assertEqual(end_count[0], 2)


class ConfigurableLeafState(State):
    def __init__(self):
        super().__init__(["done"])
        self.declare_parameter("topic", "Leaf topic")
        self.configure_count = 0
        self.configured_topic = None

    def configure(self):
        self.configure_count += 1
        self.configured_topic = self.get_parameter("topic")

    def execute(self, blackboard):
        return "done"


class TestStateMachineConfigure(unittest.TestCase):
    def test_configure_only_runs_once_on_repeated_execution(self):
        leaf = ConfigurableLeafState()
        sm = StateMachine(outcomes=["done"])
        sm.declare_parameter("topic", "Parent topic", "/demo")
        sm.add_state(
            "LEAF",
            leaf,
            transitions={"done": "done"},
            parameter_mappings={"topic": "topic"},
        )

        self.assertEqual("done", sm())
        self.assertEqual("done", sm())
        self.assertEqual(1, leaf.configure_count)
        self.assertEqual("/demo", leaf.configured_topic)

    def test_nested_configure_applies_parameter_mappings_recursively(self):
        leaf = ConfigurableLeafState()
        nested_sm = StateMachine(outcomes=["done"])
        nested_sm.declare_parameter("nested_topic", "Nested topic")
        nested_sm.add_state(
            "LEAF",
            leaf,
            transitions={"done": "done"},
            parameter_mappings={"topic": "nested_topic"},
        )

        root_sm = StateMachine(outcomes=["done"])
        root_sm.declare_parameter("root_topic", "Root topic", "/root")
        root_sm.add_state(
            "NESTED",
            nested_sm,
            transitions={"done": "done"},
            parameter_mappings={"nested_topic": "root_topic"},
        )

        self.assertEqual("done", root_sm())
        self.assertEqual(1, leaf.configure_count)
        self.assertEqual("/root", leaf.configured_topic)
        self.assertEqual("/root", nested_sm.get_parameter("nested_topic"))


class BlockingCancelState(State):
    def __init__(self, enter_event: threading.Event):
        super().__init__(outcomes=["done"])
        self.enter_event = enter_event

    def execute(self, blackboard):
        self.enter_event.set()
        while not self.is_canceled():
            time.sleep(0.01)
        return "done"


class TestStateMachineCancelBehavior(unittest.TestCase):

    def test_sequential_cancel_state_can_cancel_following_state(self):
        first_entered = threading.Event()
        second_entered = threading.Event()
        blackboard = Blackboard()
        result = {}

        sm = StateMachine(outcomes=["done"])
        sm.add_state(
            "FIRST",
            BlockingCancelState(first_entered),
            transitions={"done": "SECOND"},
        )
        sm.add_state(
            "SECOND",
            BlockingCancelState(second_entered),
            transitions={"done": "done"},
        )

        def run_state_machine():
            result["outcome"] = sm(blackboard)

        thread = threading.Thread(target=run_state_machine)
        thread.start()

        self.assertTrue(first_entered.wait(timeout=2.0))
        self.assertEqual(sm.get_current_state(), "FIRST")

        sm.cancel_state()

        self.assertTrue(second_entered.wait(timeout=2.0))
        self.assertEqual(sm.get_current_state(), "SECOND")

        sm.cancel_state()
        thread.join(timeout=2.0)

        self.assertFalse(thread.is_alive())
        self.assertEqual(result.get("outcome"), "done")

    def test_cancel_state_machine_stops_execution(self):
        first_entered = threading.Event()
        second_entered = threading.Event()
        blackboard = Blackboard()
        result = {}

        sm = StateMachine(outcomes=["done"])
        sm.add_state(
            "FIRST",
            BlockingCancelState(first_entered),
            transitions={"done": "SECOND"},
        )
        sm.add_state(
            "SECOND",
            BlockingCancelState(second_entered),
            transitions={"done": "done"},
        )

        def run_state_machine():
            try:
                sm(blackboard)
            except Exception as exc:
                result["exception"] = exc

        thread = threading.Thread(target=run_state_machine)
        thread.start()

        self.assertTrue(first_entered.wait(timeout=2.0))
        self.assertEqual(sm.get_current_state(), "FIRST")

        sm.cancel_state_machine()
        thread.join(timeout=2.0)

        self.assertFalse(thread.is_alive())
        self.assertIn("State machine canceled:", str(result.get("exception")))
        self.assertFalse(second_entered.is_set())
        self.assertEqual(sm.get_current_state(), "")


if __name__ == "__main__":
    unittest.main()
